### DISCUSSION TITLE
How build app without vercel dailydotdev/apps

### DISCUSSION THREAD

**Discussion by drunkod** [dailydotdev/daily#1585] [General]
I want build app (dailydotdev/apps) on local machine.

I get error:
 
```shell
npm run build

> webapp@0.0.0 build
> next build

  ▲ Next.js 15.0.1
  - Environments: .env

   Linting and checking validity of types  .
 ⚠ The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config
 ✓ Linting and checking validity of types    
   Creating an optimized production build ...
 ○ (serwist) The service worker will not be automatically registered, please call 'window.serwist.register()' in 'componentDidMount' or 'useEffect'.
 ○ (serwist) You may also want to add '@serwist/next/typings' to your TypeScript/JavaScript configuration file at 'compilerOptions.types'.
 ✓ (serwist) Bundling the service worker script with the URL '/sw.js' and the scope '/'...
...............
 ✓ Collecting page data    
Error occurred prerendering page "/en/sources". Read more: https://nextjs.org/docs/messages/prerender-error
TypeError: Failed to parse URL from /api/graphql
    at node:internal/deps/undici/undici:13178:13
Export encountered an error on /sources: /en/sources, exiting the build.
 ⨯ Static worker exited with code: 1 and signal: null
``` 

 My next config:
 
```javascript
      rewrites: async () => {
        const rewrites: Rewrite[] = [
          {
            source: '/api/sitemaps/:path*',
            destination: `${process.env.NEXT_PUBLIC_API_URL}/sitemaps/:path*`,
          },
          {
            source: '/search',
            destination: '/search/:provider',
            has: [
              {
                type: 'query',
                key: 'provider',
              },
            ],
          },
          {
            source: '/search',
            destination: '/search/posts',
          },
        ];

        // to support GitPod environment and avoid CORS issues, we need to proxy the API requests
        if (process.env.NEXT_PUBLIC_DOMAIN === 'localhost') {
          rewrites.unshift({
            source: '/api/:path*',
            destination: `${process.env.NEXT_PUBLIC_API_URL}/:path*`,
          });
        }

        return rewrites;
      },
``` 

      
my .env file:

```shell
      
  NEXT_PUBLIC_API_URL=http://localhost:5000
      NEXT_PUBLIC_SUBS_URL=ws://localhost:5000/graphql
      NEXT_PUBLIC_DOMAIN=ocalhost
      NEXT_PUBLIC_WEBAPP_URL=/
      NEXT_PUBLIC_AUTH_URL=http://localhost
      NEXT_PUBLIC_HEIMDALL_URL=http://localhost
```

**Comment by idoshamun** [comment]
Hi @drunkod,

In order to build the app you need to have our api running locally as well

**Reply by drunkod** [reply]
Thanks for the reply, I have run all the services that are in the docker compose file.

daily-api work on localhost :5000

```shell
community/daily-api]$ npm run dev

> dev
> nodemon ./bin/cli api | pino-pretty --messageKey message --levelKey severity

[nodemon] 3.1.7
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: ts,json
[nodemon] starting `ts-node ./bin/cli.ts api`
[12:10:35.254] INFO (9545): loading features
[12:10:35.256] ERROR (9545): failed to load features
    err: {
      "type": "Error",
      "message": "Missing clientKey",
      "stack":
          Error: Missing clientKey
              at GrowthBook._refresh (/home/alex/Documents/projects/community/daily-api/node_modules/@growthbook/growthbook/src/GrowthBook.ts:376:13)
              at GrowthBook.loadFeatures (/home/alex/Documents/projects/community/daily-api/node_modules/@growthbook/growthbook/src/GrowthBook.ts:300:33)
              at loadFeatures (/home/alex/Documents/projects/community/daily-api/src/growthbook.ts:47:14)
              at app (/home/alex/Documents/projects/community/daily-api/src/index.ts:75:21)
              at processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async run (/home/alex/Documents/projects/community/daily-api/bin/cli.ts:18:19)
    }
[12:10:35.768] USERLVL (9545): @fastify/reply-from might not behave as expected when used with @fastify/multipart
[12:10:35.774] INFO (9545): Server listening at http://127.0.0.1:5000
[12:10:35.774] INFO (9545): Server listening at http://192.168.33.212:5000
[12:10:35.774] INFO (9545): Server listening at http://172.19.0.1:5000
``` 
with config:

```shell
PORT=5000
TZ=UTC

ACCESS_SECRET='topsecret'

SCRAPER_URL=http://localhost:5001

DEFAULT_IMAGE_URL=https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/1,https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/2,https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/3,https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/4,https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/5,https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/6,https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/7,https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/8
DEFAULT_IMAGE_RATIO=1
DEFAULT_IMAGE_PLACEHOLDER=data:image/jpeg;base64,/9j/2wBDAAYEBQYFBAYGBQYHBwYIChAKCgkJChQODwwQFxQYGBcUFhYaHSUfGhsjHBYWICwgIyYnKSopGR8tMC0oMCUoKSj/2wBDAQcHBwoIChMKChMoGhYaKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCj/wAARCAAKAAoDASIAAhEBAxEB/8QAFwAAAwEAAAAAAAAAAAAAAAAABAUGB//EACYQAAIABAQGAwAAAAAAAAAAAAECAAMEBRESE0IGByExQVFScZH/xAAVAQEBAAAAAAAAAAAAAAAAAAABA//EABURAQEAAAAAAAAAAAAAAAAAAAAR/9oADAMBAAIRAxEAPwCgPMKtsdvWjpamiGsuCBVZzn3NmOAB+wYUni23kkz71OM09XObd5jBKufNagklprk4jux9QBqP82/YpQ//2Q==

URL_PREFIX=http://localhost:4000
COMMENTS_PREFIX=http://localhost:5002

COOKIES_KEY=topsecret
JWT_SECRET='|r+.2!!!.Qf_-|63*%.D'
JWT_AUDIENCE='Daily Staging'
JWT_ISSUER='Daily API Staging'

HASHNODE_WEBHOOK=http://localhost:6000/hashnode

ENABLE_SUBSCRIPTIONS=true
ENABLE_PRIVATE_ROUTES=true

REDIS_HOST=127.0.0.1
REDIS_PORT=6379

ENABLE_SETTINGS_EXPERIMENT=

INTERNAL_FEED=http://localhost:6000/feed.json
SNOTRA_ORIGIN=http://localhost:6001
PERSONALIZED_DIGEST_FEED=http://localhost:6000/feed.json
POPULAR_FEED=http://localhost:6000/popular
LOFN_ORIGIN=http://localhost:6002

POST_SCRAPER_ORIGIN=http://localhost:8000
HEIMDALL_ORIGIN=http://localhost:7000
KRATOS_ORIGIN=http://localhost:7000
MAGNI_ORIGIN=http://localhost:9000
MEILI_ORIGIN=http://localhost:7700/
SUBMIT_ARTICLE_THRESHOLD=250
ANALYTICS_URL=http://localhost:5000

MEILI_INDEX=dev
MEILI_TOKEN=dev

EXPERIMENTATION_KEY=6FA4C8D3DD2400B9D614BB7A229FD69B
#GROWTHBOOK_CLIENT_KEY=override_me_in_env.development
#GROWTHBOOK_API_CONFIG_CLIENT_KEY=override_me_in_env.development
API_CONFIG_FEATURE_KEY=api-config

JWT_PUBLIC_KEY_PATH=../apps/.cert/public.pem
JWT_PRIVATE_KEY_PATH=../apps/.cert/key.pem

# OpenTelemetry Protocol
OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4317
OTEL_TRACES_SAMPLER=parentbased_traceidratio
OTEL_TRACES_SAMPLER_ARG=1

DIGEST_QUEUE_CONCURRENCY=1

ROASTER_URL=http://localhost:7000
OG_URL=http://localhost:3000

CIO_WEBHOOK_SECRET=topsecret
CIO_REPORTING_WEBHOOK_SECRET=topsecret2

VORDR_IPS=192.0.2.0/24,198.51.100.0/24,203.0.113.0/24
VORDR_WORDS=VordrWillCatchYou,AndVordrWillHaveFun

SLACK_CLIENT_ID=clientid
SLACK_CLIENT_SECRET=topsecret
SLACK_SIGNING_SECRET=topsecret
SLACK_DB_KEY=438c104d85cecd56b88bd58d3a51d5bc

IGNORED_WORK_EMAIL_DOMAINS=igored.com,ignored.org

``` 



and others services work in docker:

```shell
daily-api-search-1 (ec08f57eacff)
Image
getmeili/meilisearch:v1.9.0 (7af9cd898077)

Ports
7700 ➔ 7700 tcp

daily-api-db-1 (7b9611c0f25d)
Image
postgres:15.4-alpine (d43df62c9cb9)

Ports
5432 ➔ 5432 tcp

daily-api-redis-1 (5079400d54a6)
Image
redis:6 (fe043551c1d6)

Ports
6379 ➔ 6379 tcp
```

**Reply by idoshamun** [reply]
@sshanzel can you have a look please?

**Reply by sshanzel** [reply]
This is similar to the issue we have in GitPod, where server-side requests fail. I will dedicate more time on this issue next week 👀

**Comment by instawirelessa** [comment]
Building an app without using platforms like Vercel for deployment is straightforward, especially if you manage your own hosting and deployment process. Here’s a step-by-step guide for [building and deploying an app](https://max.services/services/app-development/) without relying on platforms like Vercel.
Step 1: Choose Your Tech Stack
1.	Frontend Framework: Use a framework or library like:
•	React.js, Vue.js, Angular, or Svelte for web apps.
•	Plain HTML/CSS/JavaScript for simpler apps.
2.	Backend Framework: Select a backend language/framework such as:
•	Node.js (Express.js, Nest.js)
•	Python (Django, Flask)
•	Ruby (Ruby on Rails)
•	PHP (Laravel)
3.	Database: Use a database for data storage:
•	Relational: MySQL, PostgreSQL
•	NoSQL: MongoDB, Firebase Firestore
Step 2: Build Your Application
4.	Frontend Development:
•	Set up your project using a CLI tool (e.g., npx create-react-app, vue create, etc.).
•	Develop UI components and logic.
5.	Backend Development:
•	Create API endpoints using your chosen backend framework.
•	Implement authentication, business logic, and database interactions.
6.	Integration:
•	Connect your frontend with your backend API using HTTP calls (e.g., using fetch or Axios).
Step 3: Test Locally
Test your app locally on your development machine:
•	Frontend: Use tools like npm start or yarn dev.
•	Backend: Use nodemon for live reloading.
Test API integration, user workflows, and database functionality.
Step 4: Prepare for Deployment
Frontend:
Build your frontend for production using commands like:
•	bash
•	npm run build
This creates a dist or build directory containing optimized static files.
Backend:
Ensure your backend is configured for production. Set up:
•	Environment variables.
•	Secure API keys and tokens.
Database:
Migrate your database schema to the production environment using tools like Sequelize, Django Migrations, or native database tools.
Step 5: Deploy Without Vercel
Choose a Hosting Provider:
Frontend: Host static files on services like:
•	Netlify
•	AWS S3 with CloudFront
•	GitHub Pages
•	DigitalOcean Spaces
Backend: Use a cloud provider or VPS:
•	AWS EC2, DigitalOcean Droplets, or Linode
•	Google Cloud Platform or Microsoft Azure
•	Heroku (for simplified deployment)
Set Up Your Server:
Use a web server like Nginx or Apache to serve your frontend files and proxy API requests to the backend.
Configure CI/CD Pipelines (Optional):
Automate builds and deployments with GitHub Actions, GitLab CI/CD, or Jenkins.
DNS Configuration:
•	Purchase a domain and link it to your hosting using DNS settings.
•	Use services like Cloudflare for DNS management and SSL.
Step 6: Test Your Deployment
Ensure your app works in the production environment by testing:
All routes and endpoints.
Performance under load (use tools like Apache JMeter).
Security (verify HTTPS, authentication, etc.).
Step 7: Maintain and Scale
•	Monitor your app using tools like:
•	Google Cloud Monitoring, New Relic, or Sentry for error tracking.
•	Regularly update and optimize:
•	Apply security patches and update dependencies.
By taking control of hosting and deployment, you have complete flexibility and avoid platform-specific constraints. Once your app is live, you may need to scale it to handle more traffic or add new features. Here’s how to ensure your app can grow seamlessly:
Step 8: Optimize for Performance
Frontend Optimization:
•	Minimize CSS, JavaScript, and image files.
•	Use a Content Delivery Network (CDN) like Cloudflare or AWS CloudFront to cache and distribute your assets globally.
•	Enable lazy loading for images and heavy components.
Backend Optimization:
•	Use caching strategies (e.g., Redis or Memcached) to reduce database load.
•	Optimize database queries and add indexing.
•	To protect APIs from abuse Implement rate limiting.
Database Optimization:
•	Use replication for read-heavy workloads.
•	Partition or shard large datasets as your user base grows.
•	Regularly back up your database.
Step 9: Enhance Security
Frontend Security:
•	Enforce HTTPS using SSL/TLS certificates.
•	To prevent XSS attacks Validate all user inputs.
Backend Security:
•	Use strong authentication methods (e.g., OAuth2, JWTs).
•	Regularly update dependencies to fix vulnerabilities.
•	Implement security headers (e.g., Content Security Policy, X-Frame-Options).
Database Security:
•	Limit access using roles and permissions.
•	Encrypt sensitive data (e.g., passwords using bcrypt).
•	To prevent SQL injection Use parameterized queries.
Step 10: Monitor and Maintain
Monitoring:
•	Use monitoring tools like Prometheus, Grafana, or AWS CloudWatch to track performance metrics (CPU, memory, uptime).
•	Log errors and exceptions with tools like Sentry or LogRocket for frontend and backend logging.
Regular Maintenance:
•	Keep your app dependencies up to date.
•	Periodically review your codebase to refactor and optimize.
User Feedback:
Gather feedback from your users through surveys or feedback forms to prioritize feature updates.
Step 11: Prepare for Scaling
Frontend Scaling:
•	Use a CDN for distributing content and reducing load on your primary server.
•	Use server-side rendering (SSR) or static site generation (SSG) for better performance with frameworks like Next.js or Nuxt.js.
Backend Scaling:
•	Vertical Scaling: Increase server capacity (RAM, CPU) as your app grows.
•	Horizontal Scaling: Add more servers and load balancers to distribute traffic.
•	Containerization: Use Docker to create lightweight, portable application containers.
•	Orchestration: Use Kubernetes or AWS ECS for managing containerized apps.
Database Scaling:
•	Horizontal Scaling: Distribute your database load across multiple servers.
•	Implement read replicas for read-heavy applications.
•	Use managed database services like AWS RDS or Google Cloud SQL for scalability.
Step 12: Continuous Deployment
Automate Testing:
Set up automated testing pipelines with tools like Jest, Cypress, or Selenium to ensure app quality.
CI/CD Pipelines:
•	Use platforms like GitHub Actions, GitLab CI/CD, or Jenkins to automate deployment.
•	Automatically deploy updates to production after successful testing.
Feature Flags:
Implement feature flags to roll out new features gradually and monitor user response.
Step 13: Backup and Disaster Recovery
Data Backup:
•	Regularly back up all of the your database and files.
•	Use cloud storage solutions for offsite backups.
Disaster Recovery Plan:
•	Have a strategy to recover from unexpected outages.
•	Test failover systems periodically.
Summary
Building and deploying an app without platforms like Vercel offers greater control and flexibility but requires careful planning and execution. By selecting the right tools, optimizing your app, and implementing robust security and scaling practices, you can ensure a seamless and sustainable app experience for your users. With this approach, you're ready to build, deploy, and grow your app efficiently and independently.
